pennylane/transforms/batch_input.py

```diff
@@ -73,25 +72,33 @@ def circuit(inputs, weights):
     array([0.46230079, 0.73971315, 0.95666004, 0.5355225 , 0.66180948,
             0.44519553, 0.93874261, 0.9483197 , 0.78737918, 0.90866411])>
     """
-    parameters = tape.get_parameters(trainable_only=False)
-
     argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
 
-    non_trainable, trainable = [], []
-    for idx, param in enumerate(parameters):
-        if idx in argnum:
-            non_trainable.append(param)
-        else:
-            trainable.append(param)
+    all_parameters = tape.get_parameters(trainable_only=False)
+    argnum_params = [all_parameters[i] for i in argnum]
 
-    if len(np.unique([qml.math.shape(x)[0] for x in non_trainable])) != 1:
+    if any(qml.math.requires_grad(p) for p in argnum_params):
```

Comment:
I know I suggested this, but I noticed that the following probably is more consistent with how we obtain the parameters and treat the tape, _and_ cheaper:
```suggestion
    if any(num in tape.trainable_params for num in argnum):
```

Reply:
This is way better, thanks!